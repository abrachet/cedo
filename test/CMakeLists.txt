include(CTest)
enable_testing()

add_subdirectory(system)
add_subdirectory(unittests)
