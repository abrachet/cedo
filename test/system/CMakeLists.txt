function(add_cedo_system_test test_file cedo_file symbol)
    string(REGEX REPLACE "\.c(|pp)$" ".s" cedo_out ${cedo_file})
    set(cedo_out ${CMAKE_CURRENT_BINARY_DIR}/${cedo_out})

    string(REGEX REPLACE "\.c(|pp)$" ".o" cedo_input ${cedo_file})
    set(cedo_input ${CMAKE_CURRENT_BINARY_DIR}/${cedo_input})

    set(compiler ${CMAKE_CXX_COMPILER})
    if (${cedo_file} MATCHES "\.c$")
        set(compiler ${CMAKE_C_COMPILER})
    endif()
    add_custom_command(
        OUTPUT ${cedo_input}
        COMMAND ${compiler} ${CMAKE_CURRENT_SOURCE_DIR}/${cedo_file} -g -shared -fPIC -o ${cedo_input}
    )

    add_custom_command(
        OUTPUT ${cedo_out}
        DEPENDS cedo ${cedo_input}
        COMMAND ${CMAKE_BINARY_DIR}/bin/cedo -S -s ${symbol} -o ${cedo_out} ${cedo_input}
    )

    string(REGEX REPLACE "\.c(|pp)$" "" exec_name ${test_file})
    add_executable(${exec_name}
        ${cedo_out}
        ${test_file}
    )
    add_test(NAME system.${exec_name} COMMAND ${exec_name})
endfunction()

add_subdirectory(array)
add_subdirectory(compound)
add_subdirectory(typedef)

add_cedo_system_test(basic_test.c basic_test.cedo.c a)
add_cedo_system_test(main_executed_test.c main_executed_test.cedo.c a)
