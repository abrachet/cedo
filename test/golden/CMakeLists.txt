function(add_cedo_golden_test)
    cmake_parse_arguments(
        "GOLDEN"
        ""
        "CEDO_SRC;EXPECTED_OUTPUT"
        "SYMS"
        ${ARGN}
    )

    string(REGEX REPLACE "\.c(|pp)$" ".out.s" cedo_out ${GOLDEN_CEDO_SRC})
    set(cedo_out ${CMAKE_CURRENT_BINARY_DIR}/${cedo_out})

    string(REGEX REPLACE "\.c(|pp)$" ".o" cedo_input ${GOLDEN_CEDO_SRC})
    set(cedo_input ${CMAKE_CURRENT_BINARY_DIR}/${cedo_input})

    set(compiler ${CMAKE_CXX_COMPILER})
    if (${GOLDEN_CEDO_SRC} MATCHES "\.c$")
        set(compiler ${CMAKE_C_COMPILER})
    endif()
    execute_process(
        COMMAND ${compiler} ${CMAKE_CURRENT_SOURCE_DIR}/${GOLDEN_CEDO_SRC} -g -shared -fPIC -o ${cedo_input}
    )

    set(sym_list "")
    foreach(sym ${GOLDEN_SYMS})
        set(sym_list "${sym_list} -s ${sym}")
    endforeach()

    execute_process(
        COMMAND sh -c  "${CMAKE_BINARY_DIR}/bin/cedo -S ${sym_list} --no-version -o ${cedo_out} ${cedo_input}"
    )

    string(REGEX REPLACE "\.c(|pp)$" "" test_name ${GOLDEN_CEDO_SRC})
    add_test(NAME golden.${test_name} COMMAND ${CMAKE_COMMAND} -E compare_files ${cedo_out} ${CMAKE_CURRENT_SOURCE_DIR}/${GOLDEN_EXPECTED_OUTPUT})
endfunction()

add_subdirectory(compound)
add_subdirectory(pointer)

add_cedo_golden_test(CEDO_SRC basic.c EXPECTED_OUTPUT basic.s SYMS a b)
add_cedo_golden_test(CEDO_SRC nullptr.c EXPECTED_OUTPUT nullptr.s SYMS a)
